[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/ValeryiaPilchuk/Assignment-3-EquationofaSlime/blob/master/Assignment%203%20-%20Valeryia%20Pilchuk.ipynb)

In [140]:
# Imports section

import pandas as pd
import sklearn
from sklearn.model_selection import train_test_split
import numpy as np
from sklearn import svm
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures


## Part 1. Loading the dataset

In [101]:
# Using pandas load the dataset (load remotely, not locally)
url = "https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv"
data = pd.read_csv(url)
# Output the first 15 rows of the data
print(data.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(data.info)
data.describe()

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<bound method DataFrame.info of      Temperature °C  Mols KCL     Size nm^3
0               469       647  6.244743e+05
1               403       694  5.779610e+05
2               302       975  6.196847e+05
3               779       916  1.460449e+06
4               901        18  4.325726e+04
..              

Temperature °C     Mols KCL     Size nm^3
count     1000.000000  1000.000000  1.000000e+03
mean       500.500000   471.530000  5.086111e+05
std        288.819436   288.482872  4.474838e+05
min          1.000000     1.000000  1.611429e+01
25%        250.750000   226.750000  1.298267e+05
50%        500.500000   459.500000  3.827182e+05
75%        750.250000   710.250000  7.603211e+05
max       1000.000000  1000.000000  1.972127e+06

## Part 2. Splitting the dataset

In [48]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = data[["Temperature °C", "Mols KCL"]]
y = data["Size nm^3"]
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.10, random_state=42)

## Part 3. Perform a Linear Regression

In [77]:
# Use sklearn to train a model on the training set
reg = LinearRegression().fit(X, y)
# Create a sample datapoint and predict the output of that sample with the trained model
print("Sample datapoint 294, 776: \n", reg.predict(np.array([[294, 776]])))
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
print("Score: \n", reg.score(X, y))
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
print("Coefficients: \n", reg.coef_)
print("Intercept: \n", reg.intercept_)

Sample datapoint 294, 776: 
 [641825.44020401]
Score: 
 0.8607060208080753
Coefficients: 
 [ 875.90992708 1031.59502452]
Intercept: 
 -416209.8173862048


C:\Users\valer\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\LocalCache\local-packages\Python37\site-packages\sklearn\base.py:451: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  "X does not have valid feature names, but"


### Report on the score for that model, in your own words (markdown, not code) explain what the score means

The score for the model is 0.86071. This score is "the coefficient of determination of the prediction". In other words, the score is the accuracy of our linear model based on the data we have provided it.

### Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX

h(x) = 875.90992$x_1$ + 1031.59502$x_2$ - 416209.81739


## Part 4. Use Cross Validation

In [149]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
scores = cross_val_score(reg, X, y, cv=30)
# Report on their finding and their significance
print("The scores for 30 cv values are: \n", scores)
scores = cross_val_score(reg, X, y, cv=10)
print("The scores for 10 cv values are: \n", scores)
prediction = reg.predict(X_test)

The scores for 30 cv values are: 
 [0.75886543 0.83285827 0.84503685 0.89271751 0.87905771 0.8226239
 0.82596983 0.8860083  0.87968748 0.87808515 0.87500422 0.83763716
 0.89971607 0.84149677 0.87262014 0.82884345 0.86958288 0.7630771
 0.90621861 0.866835   0.86477267 0.85435504 0.88700702 0.83886377
 0.77366306 0.68806535 0.85564619 0.89361746 0.87811057 0.87748462]
The scores for 10 cv values are: 
 [0.81123596 0.86440978 0.87808742 0.86561069 0.87495621 0.84484397
 0.87941022 0.86349411 0.78353682 0.88686516]


Using cross validation we can see that the scores vary more as the cv value is increased. When the cv is 10, the scores vary only slightly, with only one values below 0.8 and all values below 0.9. When the cv value is increased to 30, the scores vary largely with the highest being above 0.9 and the lowest being below 0.7. The variety of the scores varies but is still acceptable as we do not want a value that is overfitted or underfitted. 

## Part 5. Using Polynomial Regression

In [154]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly_reg = PolynomialFeatures(degree=2)
X_poly = poly_reg.fit_transform(X)
poly = poly_reg.fit(X_poly,y)
lin_reg2 = LinearRegression()
lin_reg2.fit(X_poly,y)
# Report on the metrics and output the resultant equation as you did in Part 3.
print("Coefficients: \n", lin_reg2.coef_)
print("Intercept: \n", lin_reg2.intercept_)
scores = cross_val_score(lin_reg2, X, y, cv=30)
scores

Coefficients: 
 [ 0.00000000e+00  1.20000000e+01 -1.23111539e-07 -1.05648823e-11
  2.00000000e+00  2.85714287e-02]
Intercept: 
 1.6572012100368738e-05


array([0.75886543, 0.83285827, 0.84503685, 0.89271751, 0.87905771,
       0.8226239 , 0.82596983, 0.8860083 , 0.87968748, 0.87808515,
       0.87500422, 0.83763716, 0.89971607, 0.84149677, 0.87262014,
       0.82884345, 0.86958288, 0.7630771 , 0.90621861, 0.866835  ,
       0.86477267, 0.85435504, 0.88700702, 0.83886377, 0.77366306,
       0.68806535, 0.85564619, 0.89361746, 0.87811057, 0.87748462])

### Metrics of the model

The cross value scores are very similar to the first linear regression model that we have made, this means that both of the models create a similar outcome for test data with similar accuracy. There are 6 coeffients in this model instead of 2 which set up the equation for the polynomial regression. The intercept for this model is also a lot lower than the interept for the linear regression model because of the general shape of this model. 

### Equation for the polynomial regression

y = 12$x_1$ -1.23111$^-7$$x_2$$^2$ - 1.05649$^-11$$x_3$$^3$ + 2$x_4$$^4$ + 2.85714$^-2$$x_5$$^5$ 
